File: packages/core/tests/runtime.mjs
import test from "node:test"
import assert from "node:assert/strict"
import "../dist/node.js"
import { Authority, AuthorityCheckpoint, Change, CollaError, Document, History, HistoryCheckpoint, Ref, ServerMessage, SessionCheckpoint, Submission, SyncSession, SyncSnapshot, Value, apply, compose, invert, ref, richText, text, transform } from "../dist/node.js"

const code = expected => error => error instanceof CollaError && error.code === expected
test("Text operations reject unknown and missing types before applying a change", () => {
  const base = Value.fromJS(text("abc"))
  for (const type of ["typo", "DELETE", "", undefined, null, 2]) {
    assert.throws(() => Change.create([
      { type: "text", target: base.id, operations: [{ type, length: 1 }] },
    ]), code("invalid_argument"))
  }
  const change = Change.create([{ type: "text", target: base.id, operations: [
    { type: "retain", length: 1 },
    { type: "delete", length: 1 },
    { type: "insert", text: "!" },
  ] }])
  const next = apply(base, change)
  assert.equal(next.toJS().value, "a!c")
  assert.equal(next.id, base.id)
  assert.equal(base.toJS().value, "abc")
})

function accept(authority, session) {
  const result = authority.accept(session.outbound())
  assert.equal(result.message.type, "commit", result.message.reason?.message)
  return result
}

test("complete editing, stable identities, Move steps, inverse and immutable snapshots", () => {
  const doc = Document.create({ blocks: [{ title: text("Draft"), body: richText([]) }], archived: [], featured: null, views: 0n })
  const history = History.attach(doc)
  assert.equal(History.attach(doc), history)
  const blockId = doc.idAt(["blocks", 0])
  const before = doc.snapshot()
  let escaped, escapedText, escapedRich, escapedList
  const edit = doc.edit(tx => {
    escaped = tx
    escapedText = tx.text(["blocks", 0, "title"])
    escapedRich = tx.richText(["blocks", 0, "body"])
    escapedList = tx.list(["archived"])
    escapedText.insert(5, " v2")
    escapedRich.insertText(0, "Hello")
    escapedRich.format(0, 5, { bold: true })
    escapedRich.insertEmbed(5, ref(blockId))
    tx.set(["featured"], ref(blockId))
    tx.move(blockId, { parent: ["archived"], index: 0 })
    tx.increment(["views"], 1n)
  })
  assert.deepEqual(doc.pathOf(blockId), ["archived", 0])
  assert.equal(doc.resolve(ref(blockId)).get(["title"]).toJS().value, "Draft v2")
  assert.equal(edit.editSteps.find(step => step.type === "move").target, blockId)
  assert.ok(apply(before, edit.change).equals(edit.after))
  assert.ok(apply(before, Change.create(edit.editSteps)).equals(edit.after))
  assert.ok(apply(edit.after, invert(before, edit.change)).equals(before))
  assert.ok(Value.decode(edit.after.encode()).equals(edit.after))
  for (const fn of [() => escaped.set(["views"], 9n), () => escaped.snapshot(), () => escapedText.delete(0, 1), () => escapedRich.insertText(0, "bad"), () => escapedList.insert(0, [null])]) assert.throws(fn, code("invalid_state"))
  assert.equal(doc.get(["views"]).toJS(), 1n)
  history.undo()
  assert.ok(doc.snapshot().equals(before))
  history.redo()
  assert.ok(doc.snapshot().equals(edit.after))
  doc.close(); history.close(); doc.close()
  assert.ok(edit.before.equals(before))
  assert.ok(Value.decode(edit.after.encode()).equals(edit.after))
  assert.equal("dispose" in edit.after, false)
  assert.equal("free" in edit.after, false)
})

test("UTF-16 coordinates use each working step and reject surrogate splits", () => {
  const doc = Document.create({ title: text("A😀B"), list: [0n] })
  const original = doc.snapshot()
  assert.throws(() => doc.edit(tx => tx.text(["title"]).insert(2, "bad")), code("invalid_utf16_boundary"))
  assert.ok(doc.snapshot().equals(original))
  doc.edit(tx => { const editor = tx.text(["title"]); editor.insert(3, "!"); editor.replace(1, 2, "🐬"); editor.delete(3, 1) })
  assert.equal(doc.get(["title"]).toJS().value, "A🐬B")
  assert.throws(() => doc.edit(tx => tx.text(["title"]).delete(99, 1)), code("out_of_bounds"))
  assert.throws(() => doc.edit(tx => tx.list(["list"]).replace(0, 2, [])), code("out_of_bounds"))
  assert.throws(() => doc.edit(tx => tx.delete([])), code("invalid_argument"))
  assert.throws(() => doc.edit(tx => tx.move([], { parent: ["list"], index: 0 })), code("invalid_argument"))
})

test("Noop, thenable and nested transactions cannot change version or pending", async () => {
  const authority = Authority.create({ documentId: "atomic", value: { count: 0n, title: text("") } })
  const session = SyncSession.create({ clientId: "a", snapshot: authority.snapshot() })
  const doc = session.document
  const history = History.attach(doc)
  let events = 0
  doc.subscribe(() => events++)
  const initial = session.checkpoint().encode()
  assert.equal(doc.edit(tx => tx.increment(["count"], 0n)), null)
  assert.throws(() => doc.edit(tx => { tx.increment(["count"], 1n); doc.edit(() => {}) }), code("invalid_state"))
  assert.throws(() => doc.edit(tx => { tx.increment(["count"], 1n); return { then() {} } }), code("invalid_state"))
  let escaped
  let thenableCalls = 0
  assert.throws(() => doc.edit(tx => { tx.increment(["count"], 1n); return { then() { thenableCalls++ } } }), code("invalid_state"))
  assert.throws(() => doc.edit(async tx => { escaped = tx; tx.increment(["count"], 1n); await Promise.resolve(); tx.increment(["count"], 1n) }), code("invalid_state"))
  await new Promise(resolve => setImmediate(resolve))
  assert.equal(thenableCalls, 0)
  assert.throws(() => escaped.increment(["count"], 1n), code("invalid_state"))
  assert.throws(() => doc.edit(tx => { tx.increment(["count"], 1n); doc.close() }), code("invalid_state"))
  assert.deepEqual(session.checkpoint().encode(), initial)
  assert.equal(doc.version, 0n); assert.equal(events, 0); assert.equal(history.canUndo, false); assert.equal(session.outbound(), null)
})

test("listener failures are isolated and both event channels reject reentry", () => {
  const authority = Authority.create({ documentId: "events", value: { count: 0n } })
  const session = SyncSession.create({ clientId: "a", snapshot: authority.snapshot() })
  const doc = session.document
  let diagnostics = 0, later = 0, syncEvents = 0
  doc.subscribe(() => { throw new Error("listener") }, { onError: () => diagnostics++ })
  doc.subscribe(event => {
    assert.ok(Object.isFrozen(event)); assert.equal(doc.get(["count"]).toJS(), 1n)
    assert.throws(() => doc.edit(tx => tx.increment(["count"], 1n)), code("invalid_state")); later++
  })
  const unsubscribe = session.subscribe(state => { assert.ok(Object.isFrozen(state)); assert.throws(() => doc.edit(() => {}), code("invalid_state")); syncEvents++ })
  doc.edit(tx => tx.increment(["count"], 1n))
  assert.equal(diagnostics, 1); assert.equal(later, 1); assert.equal(syncEvents, 1)
  const result = accept(authority, session); session.receive(result.message)
  assert.equal(later, 1); assert.equal(syncEvents, 2)
  unsubscribe(); unsubscribe()
  session.receive(result.message)
  assert.equal(syncEvents, 2)
})

test("Copy/set remap internal Ref targets and Undo restores dangling references", () => {
  const source = Document.create({ target: 1n, link: null })
  source.edit(tx => tx.set(["link"], ref(source.idAt(["target"]))))
  const doc = Document.create({ item: null, copied: null })
  doc.edit(tx => tx.set(["item"], source.snapshot()))
  const itemRoot = doc.idAt(["item"]), target = doc.idAt(["item", "target"])
  const link = doc.get(["item", "link"]).toJS()
  assert.ok(link instanceof Ref); assert.equal(link.target, target); assert.notEqual(target, source.idAt(["target"]))
  const snapshot = doc.snapshot()
  const history = History.attach(doc)
  doc.edit(tx => tx.delete(target))
  assert.equal(doc.resolve(link), undefined)
  assert.equal(snapshot.resolve(link).toJS(), 1n)
  history.undo()
  assert.equal(doc.resolve(link).toJS(), 1n)
  assert.equal(doc.idAt(["item"]), itemRoot)
  doc.edit(tx => { tx.delete(["copied"]); tx.copy(itemRoot, { parent: [], key: "copied" }) })
  const copied = doc.idAt(["copied", "target"])
  assert.equal(doc.get(["copied", "link"]).toJS().target, copied)
  assert.notEqual(copied, target)
})

test("controlled protocol objects, immutable retry bytes, gaps and full checkpoints", () => {
  let authority = Authority.create({ documentId: "sync", value: { count: 0n } })
  const initial = authority.snapshot()
  assert.throws(() => SyncSession.create({ clientId: "a", snapshot: { ...initial, revision: -1n } }), code("invalid_argument"))
  let a = SyncSession.create({ clientId: "a", snapshot: SyncSnapshot.decode(initial.encode()) })
  const b = SyncSession.create({ clientId: "b", snapshot: initial })
  History.attach(a.document)
  a.document.edit(tx => tx.increment(["count"], 1n))
  const original = a.outbound(), originalBytes = original.encode()
  const mutable = original.encode(); mutable.fill(0)
  a.document.edit(tx => tx.increment(["count"], 2n))
  assert.deepEqual(a.outbound().encode(), originalBytes)
  b.document.edit(tx => tx.increment(["count"], 10n))
  let result = accept(authority, b); authority = result.authority
  const remote = ServerMessage.decode(result.message.encode())
  const before = a.checkpoint().encode()
  assert.throws(() => a.document.edit(() => a.receive(remote)), code("invalid_state"))
  assert.deepEqual(a.checkpoint().encode(), before)
  a.receive(remote); b.receive(remote)
  assert.deepEqual(a.outbound().encode(), originalBytes)
  const checkpoint = a.checkpoint().encode()
  const oldDocument = a.document
  a.close()
  assert.throws(() => oldDocument.edit(tx => tx.increment(["count"], 1n)), code("invalid_state"))
  a = SyncSession.restore(SessionCheckpoint.decode(checkpoint))
  assert.deepEqual(a.outbound().encode(), originalBytes)
  result = accept(authority, a); authority = result.authority
  const duplicate = authority.accept(Submission.decode(originalBytes))
  assert.equal(duplicate.authority.revision, authority.revision)
  assert.deepEqual(duplicate.message.encode(), result.message.encode())
  a.receive(result.message); b.receive(result.message)
  assert.equal(a.document.get(["count"]).toJS(), 13n)
  assert.equal(a.outbound().sequence, 2n)
  result = accept(authority, a); authority = result.authority
  a.receive(result.message); b.receive(result.message)
  assert.equal(a.outbound(), null)
  assert.ok(a.document.snapshot().equals(authority.snapshot().value))
  authority = Authority.restore(AuthorityCheckpoint.decode(authority.checkpoint().encode()))
  assert.ok(authority.snapshot().value.equals(b.document.snapshot()))
  const late = SyncSession.create({ clientId: "late", snapshot: initial })
  const lateBefore = late.checkpoint().encode()
  assert.throws(() => late.receive(result.message), code("missing_revision"))
  assert.deepEqual(late.checkpoint().encode(), lateBefore)
  for (const message of authority.commitsSince(0n)) late.receive(message)
  assert.ok(late.document.snapshot().equals(authority.snapshot().value))
})

test("session checkpoints larger than 64 MiB restore with the original pending request", () => {
  const mib = 1024 * 1024
  const authority = Authority.create({ documentId: "large", value: {
    first: "x".repeat(12 * mib), second: "y".repeat(12 * mib), count: 0n,
  } })
  const session = SyncSession.create({ clientId: "a", snapshot: authority.snapshot() })
  let restored
  try {
    assert.ok(session.checkpoint().encode().length < 64 * mib)
    session.document.edit(tx => tx.increment(["count"], 1n))
    const pending = session.outbound().encode()
    const saved = session.checkpoint().encode()
    assert.ok(saved.length > 64 * mib)
    session.close()
    restored = SyncSession.restore(SessionCheckpoint.decode(saved))
    assert.deepEqual(restored.outbound().encode(), pending)
    assert.equal(restored.document.get(["count"]).toJS(), 1n)
    const result = accept(authority, restored)
    restored.receive(result.message)
    assert.equal(restored.outbound(), null)
    assert.ok(restored.document.snapshot().equals(result.authority.snapshot().value))
  } finally {
    restored?.close()
    session.close()
  }
})

test("grouped history checkpoints and public transform return order", () => {
  const doc = Document.create({ title: text("a") })
  const history = History.attach(doc)
  const before = doc.snapshot()
  doc.edit(tx => tx.text(["title"]).insert(1, "b"), { group: "typing" })
  doc.edit(tx => tx.text(["title"]).insert(2, "c"), { group: "typing" })
  const restored = Document.create(doc.snapshot())
  const restoredHistory = History.restore(restored, HistoryCheckpoint.decode(history.checkpoint().encode()))
  restoredHistory.undo(); assert.ok(restored.snapshot().equals(before)); assert.equal(restoredHistory.canUndo, false)
  restoredHistory.redo(); assert.ok(restored.snapshot().equals(doc.snapshot()))
  const leftDoc = Document.create(before), rightDoc = Document.create(before)
  const left = leftDoc.edit(tx => tx.text(["title"]).insert(1, "L")).change
  const right = rightDoc.edit(tx => tx.text(["title"]).insert(1, "R")).change
  const [leftAfterRight, rightAfterLeft] = transform(before, left, right, { priority: "left" })
  assert.ok(apply(apply(before, right), leftAfterRight).equals(apply(apply(before, left), rightAfterLeft)))
  assert.ok(apply(before, compose(before, left, rightAfterLeft)).equals(apply(apply(before, left), rightAfterLeft)))
})

test("version-2 codecs reject other types, old versions and trailing bytes", () => {
  const value = Value.fromJS({ n: 1n, title: text("😀") })
  const encoded = value.encode()
  assert.throws(() => Change.decode(encoded), code("invalid_encoding"))
  const old = encoded.slice(); old[5] = 1
  assert.throws(() => Value.decode(old), code("invalid_encoding"))
  const trailing = new Uint8Array(encoded.length + 1); trailing.set(encoded)
  assert.throws(() => Value.decode(trailing), code("invalid_encoding"))
  assert.throws(() => Value.fromJS(1n << 64n), code("integer_overflow"))
  assert.throws(() => Value.fromJS(Number.NaN), code("invalid_value"))
  assert.throws(() => Value.fromJS("\ud800"), code("invalid_value"))
  const cyclic = {}; cyclic.self = cyclic
  assert.throws(() => Value.fromJS(cyclic), code("invalid_value"))
  assert.ok(value.equals(Value.decode(encoded)))
})

test("one-hop cyclic Ref and canonical formatting floats", () => {
  const doc = Document.create({ a: null, b: null, body: richText([{ type: "text", text: "x", attrs: { scale: -0 } }]) })
  const a = doc.idAt(["a"]), b = doc.idAt(["b"])
  doc.edit(tx => { tx.set(a, ref(b)); tx.set(b, ref(a)) })
  assert.equal(doc.resolve(ref(a)).toJS().target, b)
  assert.equal(doc.resolve(ref(b)).toJS().target, a)
  assert.ok(doc.snapshot().equals(Value.decode(doc.snapshot().encode())))
  assert.equal(Object.is(doc.get(["body"]).toJS().spans[0].attrs.scale, -0), false)
  assert.throws(() => doc.edit(tx => tx.increment(a, 0n)), code("type_mismatch"))
})

File: crates/colla/tests/runtime.rs
use colla::{
    Authority, AuthorityCheckpoint, Body, CollaError, Document, ErrorCode, History,
    HistoryCheckpoint, Segment, ServerMessage, SessionCheckpoint, SyncSession, Value,
};

fn map(values: impl IntoIterator<Item = (&'static str, Value)>) -> Value {
    Value::map(values.into_iter().map(|(k, v)| (k.into(), v))).unwrap()
}
fn publish(authority: &mut Authority, session: &SyncSession) -> ServerMessage {
    let request = session.outbound().unwrap().unwrap();
    let (next, message) = authority.accept(&request).unwrap();
    *authority = next;
    assert!(matches!(message, ServerMessage::Commit(_)), "{message:?}");
    message
}

#[test]
fn failed_and_nested_edits_leave_all_participants_unchanged() {
    let value = map([
        ("count", Value::int(i64::MAX)),
        ("text", Value::text("A😀B").unwrap()),
    ]);
    let authority = Authority::create("doc", value).unwrap();
    let session = SyncSession::create("a", authority.snapshot()).unwrap();
    let doc = session.document();
    let history = History::attach(&doc).unwrap();
    let before = session.checkpoint().unwrap().encode();
    let result = doc.edit(|tx| {
        tx.text_replace(vec![Segment::Key("text".into())], 0, 1, "a")?;
        assert_eq!(
            doc.edit(|_| Ok(())).unwrap_err().code,
            ErrorCode::InvalidState
        );
        assert_eq!(doc.close().unwrap_err().code, ErrorCode::InvalidState);
        tx.increment(vec![Segment::Key("count".into())], 1)
    });
    assert_eq!(result.unwrap_err().code, ErrorCode::IntegerOverflow);
    assert_eq!(before, session.checkpoint().unwrap().encode());
    assert!(!history.can_undo().unwrap());
    doc.edit(|tx| {
        let path = vec![Segment::Key("text".into())];
        assert_eq!(
            tx.utf16_to_scalar(path.clone(), 2).unwrap_err().code,
            ErrorCode::InvalidUtf16Boundary
        );
        assert_eq!(tx.utf16_to_scalar(path, 3)?, 2);
        Ok(())
    })
    .unwrap();
    assert_eq!(doc.version().unwrap(), 0);
    let snapshot = doc.snapshot().unwrap();
    doc.close().unwrap();
    doc.close().unwrap();
    assert_eq!(Value::decode(&snapshot.encode()).unwrap(), snapshot);
}

#[test]
fn groups_and_checkpoints_restore_original_identities() {
    let child = Value::int(1);
    let list = Value::list(vec![child.clone()]).unwrap();
    let doc = Document::create(list.clone()).unwrap();
    let history = History::attach(&doc).unwrap();
    for _ in 0..3 {
        doc.edit_group(Some("typing".into()), |tx| tx.increment(child.id(), 1))
            .unwrap();
    }
    assert_eq!(doc.get(child.id()).unwrap().body(), &Body::Int(4));
    let checkpoint = HistoryCheckpoint::decode(&history.checkpoint().unwrap().encode()).unwrap();
    let restored = Document::create(doc.snapshot().unwrap()).unwrap();
    let restored_history = History::restore(&restored, checkpoint).unwrap();
    restored_history.undo().unwrap();
    assert_eq!(restored.snapshot().unwrap(), list);
    assert!(!restored_history.can_undo().unwrap());
    restored_history.redo().unwrap();
    assert_eq!(restored.snapshot().unwrap(), doc.snapshot().unwrap());
}

#[test]
fn immutable_retries_buffer_promotion_and_restart() {
    let value = Value::int(0);
    let id = value.id();
    let mut authority = Authority::create("doc", value).unwrap();
    let session = SyncSession::create("a", authority.snapshot()).unwrap();
    let doc = session.document();
    History::attach(&doc).unwrap();
    doc.edit(|tx| tx.increment(id, 1)).unwrap();
    let original = session.outbound().unwrap().unwrap();
    let bytes = original.encode();
    let mut external = original.encode();
    external.fill(0);
    doc.edit(|tx| tx.increment(id, 2)).unwrap();
    assert_eq!(session.outbound().unwrap().unwrap().encode(), bytes);
    let restored = SyncSession::restore(
        SessionCheckpoint::decode(&session.checkpoint().unwrap().encode()).unwrap(),
    )
    .unwrap();
    let message = publish(&mut authority, &restored);
    let revision = authority.revision();
    let (same, duplicate) = authority.accept(&original).unwrap();
    assert_eq!(same.revision(), revision);
    assert_eq!(duplicate, message);
    restored.receive(&message).unwrap();
    assert_eq!(restored.document().get(id).unwrap().body(), &Body::Int(3));
    let second = restored.outbound().unwrap().unwrap();
    assert_eq!(second.sequence(), 2);
    assert_eq!(second.base_revision(), 1);
    let message = publish(&mut authority, &restored);
    restored.receive(&message).unwrap();
    restored.receive(&message).unwrap();
    assert!(restored.outbound().unwrap().is_none());
    assert_eq!(
        restored.document().snapshot().unwrap(),
        *authority.snapshot().value()
    );
    let recovered =
        Authority::restore(AuthorityCheckpoint::decode(&authority.checkpoint().encode()).unwrap())
            .unwrap();
    assert_eq!(recovered.snapshot(), authority.snapshot());
}

#[test]
fn remote_reentry_gaps_and_recovery_preserve_work() {
    let value = Value::int(0);
    let id = value.id();
    let mut authority = Authority::create("doc", value).unwrap();
    let a = SyncSession::create("a", authority.snapshot()).unwrap();
    let b = SyncSession::create("b", authority.snapshot()).unwrap();
    b.document().edit(|tx| tx.increment(id, 1)).unwrap();
    let first = publish(&mut authority, &b);
    b.receive(&first).unwrap();
    b.document().edit(|tx| tx.increment(id, 1)).unwrap();
    let second = publish(&mut authority, &b);
    let before = a.checkpoint().unwrap().encode();
    assert_eq!(
        a.receive(&second).unwrap_err().code,
        ErrorCode::MissingRevision
    );
    assert_eq!(before, a.checkpoint().unwrap().encode());
    a.document()
        .edit(|_| {
            assert_eq!(a.receive(&first).unwrap_err().code, ErrorCode::InvalidState);
            Ok(())
        })
        .unwrap();
    assert_eq!(before, a.checkpoint().unwrap().encode());
    a.document().edit(|tx| tx.increment(id, 9)).unwrap();
    authority = authority.compact(2).unwrap();
    let (_, rejection) = authority.accept(&a.outbound().unwrap().unwrap()).unwrap();
    a.receive(&rejection).unwrap();
    assert_eq!(
        a.recovery_reason().unwrap().unwrap().code,
        ErrorCode::HistoryExpired
    );
    assert!(a.outbound().unwrap().is_none());
    a.document().edit(|tx| tx.increment(id, 1)).unwrap();
    assert_eq!(a.document().get(id).unwrap().body(), &Body::Int(10));
    SyncSession::restore(SessionCheckpoint::decode(&a.checkpoint().unwrap().encode()).unwrap())
        .unwrap();
}

#[test]
fn undo_move_keeps_remote_content_and_syncs_as_a_new_request() {
    let item = Value::int(0);
    let from = Value::list(vec![item.clone()]).unwrap();
    let to = Value::list(vec![]).unwrap();
    let mut authority =
        Authority::create("doc", map([("from", from.clone()), ("to", to.clone())])).unwrap();
    let a = SyncSession::create("a", authority.snapshot()).unwrap();
    let b = SyncSession::create("b", authority.snapshot()).unwrap();
    let history = History::attach(&a.document()).unwrap();
    a.document()
        .edit(|tx| tx.move_to(item.id(), to.id(), Segment::Index(0)))
        .unwrap();
    b.document().edit(|tx| tx.increment(item.id(), 7)).unwrap();
    let message = publish(&mut authority, &b);
    a.receive(&message).unwrap();
    b.receive(&message).unwrap();
    let message = publish(&mut authority, &a);
    a.receive(&message).unwrap();
    b.receive(&message).unwrap();
    history.undo().unwrap();
    assert_eq!(a.document().get(item.id()).unwrap().body(), &Body::Int(7));
    assert_eq!(
        a.document().snapshot().unwrap().path_of(item.id()).unwrap(),
        vec![Segment::Key("from".into()), Segment::Index(0)]
    );
    let message = publish(&mut authority, &a);
    a.receive(&message).unwrap();
    b.receive(&message).unwrap();
    assert_eq!(
        a.document().snapshot().unwrap(),
        b.document().snapshot().unwrap()
    );
}

#[test]
fn three_clients_with_delays_retries_buffers_and_restarts_converge() {
    let items: Vec<_> = (0..4).map(Value::int).collect();
    let list = Value::list(items.clone()).unwrap();
    let text = Value::text("😀").unwrap();
    let mut authority = Authority::create(
        "random",
        map([("items", list.clone()), ("title", text.clone())]),
    )
    .unwrap();
    let mut sessions: Vec<_> = (0..3)
        .map(|i| SyncSession::create(format!("c{i}"), authority.snapshot()).unwrap())
        .collect();
    for session in &sessions {
        History::attach(&session.document()).unwrap();
    }
    let mut seed = 41u64;
    for step in 0..180 {
        seed = seed.wrapping_mul(6364136223846793005).wrapping_add(1);
        let client = (seed >> 32) as usize % 3;
        let session = &sessions[client];
        session
            .document()
            .edit(|tx| match step % 3 {
                0 => tx.increment(items[step % items.len()].id(), 1),
                1 => tx.text_replace(text.id(), 0, 0, &format!("{client}")),
                _ => tx.move_to(
                    items[step % items.len()].id(),
                    list.id(),
                    Segment::Index(step % 4),
                ),
            })
            .unwrap();
        if step % 2 == 0 && session.outbound().unwrap().is_some() {
            publish(&mut authority, session);
        }
        let receiver = (client + 1) % 3;
        let commits = authority
            .commits_since(sessions[receiver].revision().unwrap())
            .unwrap();
        if let Some(commit) = commits.first() {
            let message = ServerMessage::Commit(commit.clone());
            sessions[receiver].receive(&message).unwrap();
            sessions[receiver].receive(&message).unwrap();
        }
        if step % 17 == 0 {
            sessions[client] = SyncSession::restore(
                SessionCheckpoint::decode(&sessions[client].checkpoint().unwrap().encode())
                    .unwrap(),
            )
            .unwrap();
        }
        if step % 23 == 0 {
            authority = Authority::restore(
                AuthorityCheckpoint::decode(&authority.checkpoint().encode()).unwrap(),
            )
            .unwrap();
        }
    }
    for _ in 0..30 {
        for session in &sessions {
            if session.outbound().unwrap().is_some() {
                publish(&mut authority, session);
            }
        }
        for session in &sessions {
            for commit in authority
                .commits_since(session.revision().unwrap())
                .unwrap()
            {
                session.receive(&ServerMessage::Commit(commit)).unwrap();
            }
        }
        if sessions.iter().all(|s| s.outbound().unwrap().is_none()) {
            break;
        }
    }
    for session in &sessions {
        assert!(session.outbound().unwrap().is_none());
        assert!(session.recovery_reason().unwrap().is_none());
        assert_eq!(
            session.document().snapshot().unwrap(),
            *authority.snapshot().value()
        );
    }
}

#[test]
fn explicit_callback_error_rolls_back() {
    let value = Value::int(0);
    let id = value.id();
    let doc = Document::create(value.clone()).unwrap();
    let result = doc.edit(|tx| {
        tx.increment(id, 1)?;
        Err(CollaError::new(ErrorCode::InvalidArgument, "abort"))
    });
    assert!(result.is_err());
    assert_eq!(doc.snapshot().unwrap(), value);
}

#[test]
fn undo_one_move_preserves_the_other_clients_move() {
    let a = Value::int(0);
    let b = Value::int(1);
    let c = Value::int(2);
    let value = Value::list(vec![a.clone(), b.clone(), c.clone()]).unwrap();
    let root = value.id();
    let mut authority = Authority::create("moves", value).unwrap();
    let left = SyncSession::create("left", authority.snapshot()).unwrap();
    let right = SyncSession::create("right", authority.snapshot()).unwrap();
    let history = History::attach(&left.document()).unwrap();
    left.document()
        .edit(|tx| tx.move_to(a.id(), root, Segment::Index(2)))
        .unwrap();
    right
        .document()
        .edit(|tx| tx.move_to(b.id(), root, Segment::Index(2)))
        .unwrap();
    let message = publish(&mut authority, &right);
    left.receive(&message).unwrap();
    right.receive(&message).unwrap();
    history.undo().unwrap();
    assert_eq!(
        left.document().snapshot().unwrap().body(),
        &Body::List(vec![a, c, b])
    );
}

#[test]
fn undo_rich_insertion_preserves_remote_formatting() {
    use colla::{Attr, AttrPatch, Attrs, RichOp, RichSpan};
    let value = Value::rich_text(vec![RichSpan::Text {
        text: "ab".into(),
        attrs: Attrs::new(),
    }])
    .unwrap();
    let root = value.id();
    let mut authority = Authority::create("rich", value).unwrap();
    let left = SyncSession::create("left", authority.snapshot()).unwrap();
    let right = SyncSession::create("right", authority.snapshot()).unwrap();
    let history = History::attach(&left.document()).unwrap();
    left.document()
        .edit(|tx| {
            tx.rich_text_edit(
                root,
                vec![
                    RichOp::Retain {
                        len: 1,
                        attrs: AttrPatch::new(),
                    },
                    RichOp::Insert(RichSpan::Text {
                        text: "X".into(),
                        attrs: Attrs::new(),
                    }),
                ],
            )
        })
        .unwrap();
    right
        .document()
        .edit(|tx| {
            tx.rich_text_edit(
                root,
                vec![
                    RichOp::Retain {
                        len: 1,
                        attrs: AttrPatch::new(),
                    },
                    RichOp::Retain {
                        len: 1,
                        attrs: AttrPatch::from([("bold".into(), Some(Attr::Bool(true)))]),
                    },
                ],
            )
        })
        .unwrap();
    let expected = right.document().snapshot().unwrap();
    let message = publish(&mut authority, &right);
    left.receive(&message).unwrap();
    right.receive(&message).unwrap();
    history.undo().unwrap();
    assert_eq!(left.document().snapshot().unwrap(), expected);
}

#[test]
fn history_restores_deleted_and_copied_identities_and_skips_remote_noops() {
    let value = Value::int(1);
    let original = value.id();
    let root = map([
        ("item", value),
        ("list", Value::list(vec![]).unwrap()),
        ("link", Value::reference(original)),
    ]);
    let doc = Document::create(root.clone()).unwrap();
    let history = History::attach_with_capacity(&doc, 2).unwrap();
    doc.edit(|tx| tx.delete(original)).unwrap();
    assert!(doc
        .resolve(colla::Ref { target: original })
        .unwrap()
        .is_none());
    history.undo().unwrap();
    assert_eq!(doc.snapshot().unwrap(), root);
    let mut copied = None;
    doc.edit(|tx| {
        copied = Some(tx.copy(
            original,
            vec![Segment::Key("list".into())],
            Segment::Index(0),
        )?);
        Ok(())
    })
    .unwrap();
    let copied = copied.unwrap();
    history.undo().unwrap();
    assert!(!doc.has(copied).unwrap());
    history.redo().unwrap();
    assert!(doc.has(copied).unwrap());
    assert_eq!(doc.references_to(original).unwrap().len(), 1);
    let authority = Authority::create("history", root).unwrap();
    let a = SyncSession::create("a", authority.snapshot()).unwrap();
    let b = SyncSession::create("b", authority.snapshot()).unwrap();
    let history = History::attach(&a.document()).unwrap();
    a.document().edit(|tx| tx.increment(original, 1)).unwrap();
    b.document().edit(|tx| tx.delete(original)).unwrap();
    let (authority, commit) = authority.accept(&b.outbound().unwrap().unwrap()).unwrap();
    a.receive(&commit).unwrap();
    assert!(!history.can_undo().unwrap());
    assert!(history.undo().unwrap().is_none());
    let (_, own) = authority.accept(&a.outbound().unwrap().unwrap()).unwrap();
    a.receive(&own).unwrap();
    assert!(!a.document().has(original).unwrap());
    assert!(a.outbound().unwrap().is_none());
}

#[test]
fn injected_allocator_burns_failed_transaction_ids() {
    let mut allocator = colla::IdAllocator::deterministic([91; 16]);
    allocator.scope(|| {
        let base = Value::int(0);
        let doc = Document::create(base.clone()).unwrap();
        let abandoned = Value::int(1);
        assert!(doc
            .edit(|tx| {
                tx.set(base.id(), abandoned.clone())?;
                Err(CollaError::new(ErrorCode::InvalidArgument, "abort"))
            })
            .is_err());
        let next = Value::int(2);
        assert_eq!(next.id().namespace(), [91; 16]);
        assert!(next.id().sequence() > abandoned.id().sequence() + 1);
        assert_eq!(doc.snapshot().unwrap(), base);
    });
    assert!(allocator.allocate().unwrap().sequence() > 4);
}
